Add mtls doc (#449)
SAP/cloud-sdk: docs/java/features/connectivity/mtls.mdx

---
id: sdk-connectivity-mtls
title: Certificate Based Authentication using SAP Cloud SDK
hide_title: false
hide_table_of_contents: false
sidebar_label: Certificate Based Authentication
description: This article describes how the SAP Cloud SDK for Java can be used to establish connections to other cloud services using mTLS
keywords:
  - sap
  - cloud
  - sdk
  - mTLS
  - java
  - connectivity
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

The SAP Cloud SDK now supports certificate based authentication while establishing connections to other cloud services.
:::info
Currently being limited to `XSUAA`, support exists but is not tested extensively yet for `Destination` and `Connectivity` services.
:::

## Why mTLS
Mutual `TLS` or `mTLS` for short is a method for mutual authentication.
It ensures that before proceeding with the `HTTP` exchange both the client and the server each mutually verify each other's identities by the use of `X.509` certificates.

> ## Usage of Client Credentials and shortcomings

Reviewer: Title case

```suggestion
## Usage of Client Credentials and Shortcomings
```

> When a service or an application is protected using [`XSUAA`](https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/6373bb7a96114d619bfdfdc6f505d1b9.html), we need to obtain an access token from `XSUAA` to access the service or application.

Reviewer: Please do not use "we" (this is done more often than just here)

```suggestion
If a service or an application is protected using [`XSUAA`](https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/6373bb7a96114d619bfdfdc6f505d1b9.html), an access token from `XSUAA` is required to access the protected service or application.
```

To obtain an access token or `JWT`, first we extract the `clientid`, `clientsecret` and `uri` from the service binding and then using the `url` from `xsuaa` service binding, trigger a [`token flow`](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/token-client) to obtain a `JWT`.

> Please note that the process of obtaining the `JWT` is internally carried out by the SAP Cloud SDK and developers need not worry about doing this explicitly.

Reviewer: ```suggestion
Please note that the process of obtaining the `JWT` is internally carried out by the SAP Cloud SDK and application developers do not need to worry about doing this explicitly.
```

> Now, adding this obtained `JWT` as an `Authorisation` header to the request to the service, ensures that your request gets authorised and authenticated.

Reviewer: The HTTP Header is usually written in American English ([reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)).
In general, I think we usually prefer American English (cc @artemkovalyov please correct me if I'm wrong here)

```suggestion
Now, adding this obtained `JWT` as an `Authorization` header to the request to the service, ensures that your request gets authorized and authenticated.
```

However there is always the inherent risk that these credentials may get leaked.
These credentials are typically not frequently changed as well.
With the use of certificates the risk of leaking powerful technical user credentials can be mitigated.

> ## Certificate based Authentication for services (Soon to be default)

Reviewer: Please check the go-live date once again.

```suggestion
## Certificate Based Authentication for Services

:::info Hint
Certificate based authentication is to be the default authentication mechanism on SAP BTP Cloud Foundry starting from Sept. 13th 2021.
:::
```

To adhere to the latest recommended security best practices by SAP, some re-use and kernel services have enabled certificate based authentication.
As a consequence, SAP Cloud SDK also supports certificate based authentication while accessing these services.
In general, if a service binding has a property `credentials: { credential-type: x509 }` then this is an indication that certificate based authentication can be used to access the service.

`XSUAA` service is one among the services that has enabled certificate based authentication.
This means that in the service binding in addition to the `client_secret` you will also find a `certificate` and `key` entry now.

> For e.g.

Reviewer: ```suggestion
Example:
```

```
{
    "VCAP_SERVICES": {
		"xsuaa": [
			{
				"label": "xsuaa",
				...
				"credentials": {
				    ...
				    "credential-type": "x509"
					"clientid": "fictional-client-id",
					"clientsecret": "fictional-secret",
					"certificate": "-----BEGIN CERTIFICATE-----fictional-certificate-----END CERTIFICATE-----\n",
					"key": "key"
					...
				},
			}
		]
	}
}
```
This certificate would now be used to establish a secure session to obtain the `JWT` instead of the client credentials token flow used previously.
Please note that the process of obtaining the `JWT` is still internally carried out by the SAP Cloud SDK, we just use the `certificate` for this process now if the binding contains the `certificate`.

## Validity of Certificates
The `X.509` certificates used in the service binding could either be external or `XSUAA` managed.

> :::caution

Reviewer 2: ```suggestion
:::danger
```

> By default `XSUAA` managed certificates are only valid for 7 days and to obtain a new certificate, deletion and re-creation of service binding is required.

Reviewer: ```suggestion
By default `XSUAA` managed certificates are valid only for 7 days.
To obtain a new certificate, deletion and re-creation of service binding is required.
```

The application also needs to be restarted to update the service binding.

:::

To extend the validity of the `XSUAA` managed certificate you could use parameters while creating a service binding.

`cf bind-service myapp myservice -c parameters.json`

```json
{
  "credential-type": "x509",
  "x509": {
    "key-length": 2048,
    "validity": 365,
    "validity-type": "DAYS"
  }
}
```
This way you can extend the validity of the generated `certificate` to a maximum of one year.
You can find more details on X.509 certificates [here](https://github.wdf.sap.corp/pages/CPSecurity/Knowledge-Base/03_ApplicationSecurity/X509Authentication/).

## Sticking to Client Credentials
If you do not want to use Certificate based Authentication to obtain your `JWT`s and still want to continue using Client Credentials you can do that by configuring your service instances.
Ensure that your `xs-security.json` contains the `instance-secret` as the first entry in property `credential-types`

```json
"oauth2-configuration": {

>   "credential-types": ["instance-secret",...]

Reviewer 2: I think if `x509` was added here in place of the `...` both would be available and the SDK would take the certificate. So maybe remove the dots.

description: This article describes how the SAP Cloud SDK for Java can be used to establish connections to other cloud services using mTLS
keywords:
  - sap
  - cloud
  - sdk
  - mTLS
  - java
  - connectivity
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::danger
Currently support for Certificate based Authentication from SAP Cloud SDK is limited to `XSUAA` service.
Support exists but is not tested extensively yet for `Destination` and `Connectivity` services.
:::

The SAP Cloud SDK now supports certificate based authentication while establishing connections to other cloud services.

## Why mTLS
Mutual `TLS` or `mTLS` for short is a method for mutual authentication.
It ensures that before proceeding with the `HTTP` exchange both the client and the server each mutually verify each other's identities by the use of `X.509` certificates.

## Usage of Client Credentials and Shortcomings
If a service or an application is protected using [`XSUAA`](https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/6373bb7a96114d619bfdfdc6f505d1b9.html), an access token from `XSUAA` is required to access the protected service or application.

> To obtain an access token or `JWT`, first we extract the `clientid`, `clientsecret` and `uri` from the service binding and then using the `url` from `xsuaa` service binding, trigger a [`token flow`](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/token-client) to obtain a `JWT`.

Reviewer: There are still some `we`s in the documentation, which is not only a little bit too informal (who is `we`? Most likely you either mean the Cloud SDK or the application itself) but also violates compliancy AFAIK as usually `we` would be associated with SAP as a company.

Author: Yes, I am still looking into changing them all.

Author: Thanks for pointing it out, I will pay more attention to this while writing documents in future.

As a consequence, SAP Cloud SDK also supports certificate based authentication while accessing these services.
In general, if a service binding has a property `credentials: { credential-type: x509 }` then this is an indication that certificate based authentication can be used to access the service.

`XSUAA` service is one among the services that has enabled certificate based authentication.
This means that in the service binding in addition to the `client_secret` you will also find a `certificate` and `key` entry now.
Example:
```
{
    "VCAP_SERVICES": {
		"xsuaa": [
			{
				"label": "xsuaa",
				...
				"credentials": {
				    ...
				    "credential-type": "x509"
					"clientid": "fictional-client-id",
					"clientsecret": "fictional-secret",
					"certificate": "-----BEGIN CERTIFICATE-----fictional-certificate-----END CERTIFICATE-----\n",
					"key": "key"
					...
				},
			}
		]
	}
}
```
This certificate would now be used to establish a secure session to obtain the `JWT` instead of the client credentials token flow used previously.

> Please note that the process of obtaining the `JWT` is still internally carried out by the SAP Cloud SDK, we just use the `certificate` for this process now if the binding contains the `certificate`.

Reviewer: Here is another `we`.

---
id: sdk-connectivity-mtls
title: Certificate Based Authentication using SAP Cloud SDK
hide_title: false
hide_table_of_contents: false
sidebar_label: Certificate Based Authentication
description: This article describes how the SAP Cloud SDK for Java can be used to establish connections to other cloud services using mTLS
keywords:
  - sap
  - cloud
  - sdk
  - mTLS
  - java
  - connectivity
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::danger
Currently support for Certificate based Authentication from SAP Cloud SDK is limited to `XSUAA` service.
Support exists but is not tested extensively yet for `Destination` and `Connectivity` services.
:::

The SAP Cloud SDK now supports certificate based authentication while establishing connections to other cloud services.

## Why mTLS
Mutual `TLS` or `mTLS` for short is a method for mutual authentication.
It ensures that before proceeding with the `HTTP` exchange both the client and the server each mutually verify each other's identities by the use of `X.509` certificates.

> ## Usage of Client Credentials and Shortcomings

Author: I will prefer the old section title because I think I don't elaborate more on `mTLS` in this section to add it to the heading.

> ## Usage of Client Credentials and Shortcomings

Author: We discussed about this and we feel the `Shortcomings of Client Credentials` is a better fit.

To obtain an access token or `JWT`, first the `clientid`, `clientsecret` and `uri` is extracted from the service binding and then using the `url` from `xsuaa` service binding, a [`token flow`](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/token-client) is triggered to obtain a `JWT`.

Please note that the process of obtaining the `JWT` is internally carried out by the SAP Cloud SDK and application developers do not need to worry about doing this explicitly.
Now, adding this obtained `JWT` as an `Authorization` header to the request to the service, ensures that your request gets authorized and authenticated.

However there is always the inherent risk that these credentials may get leaked.
These credentials are typically not frequently changed as well.
With the use of certificates the risk of leaking powerful technical user credentials can be mitigated.

## Certificate Based Authentication for Services

:::info Hint
Certificate based authentication is presumably going to be the default authentication mechanism on SAP BTP Cloud Foundry, Canary landscape starting from Sept. 13th 2021.

> :::

Author: Is it not better to have a concrete date ?

> To adhere to the latest recommended security best practices by SAP, some re-use and kernel services have enabled certificate based authentication.
> As a consequence, SAP Cloud SDK also supports certificate based authentication while accessing these services.
> In general, if a service binding has a property `credentials: { credential-type: x509 }` then this is an indication that certificate based authentication can be used to access the service.

Reviewer 3: ```suggestion
To adhere to the latest recommended security best practices by SAP, some re-use and kernel services have enabled certificate-based authentication.
As a consequence, SAP Cloud SDK also supports certificate-based authentication while accessing these services.
If a service binding has a property `credentials: { credential-type: x509 }` then this is an indication that you can use certificate-based authentication to access the service.
```

> `XSUAA` service is one among the services that has enabled certificate based authentication.
> This means that in the service binding in addition to the `client_secret` you will also find a `certificate` and `key` entry now.

Reviewer 3: ```suggestion
`XSUAA` service is one of the services that has enabled certificate-based authentication.
This means that in the service binding in addition to the `client_secret` you will also find `certificate` and `key` entries now.
```

Example:
```
{
    "VCAP_SERVICES": {
		"xsuaa": [
			{
				"label": "xsuaa",
				...
				"credentials": {
				    ...
				    "credential-type": "x509"
					"clientid": "fictional-client-id",
					"clientsecret": "fictional-secret",
					"certificate": "-----BEGIN CERTIFICATE-----fictional-certificate-----END CERTIFICATE-----\n",
					"key": "key"
					...
				},
			}
		]
	}
}
```

> This certificate would now be used to establish a secure session to obtain the `JWT` instead of the client credentials token flow used previously.
> Please note that the process of obtaining the `JWT` is still internally carried out by the SAP Cloud SDK.
> The only difference is that the provided `certificate` is used to authenticate the client.

Reviewer 3: ```suggestion
This certificate would now be used to establish a secure session to obtain the `JWT` instead of the client credentials token flow used previously.
The SAP Cloud SDK will as before fetch the `JWT` for you.
The only difference is we will use the `certificate` from your service binding to authenticate the client.
```

Author: I am just editing so that we don't have a `we` in the sentence. Johannes wrote earlier that this could be informal.

## Validity of Certificates
The `X.509` certificates used in the service binding could either be external or `XSUAA` managed.

> :::danger
> 
> By default `XSUAA` managed certificates are valid only for 7 days.
> To obtain a new certificate, deletion and re-creation of service binding is required.
> The application also needs to be restarted to update the service binding.
> 
> :::

Reviewer 3: ```suggestion
:::danger short certificate validity period

By default `XSUAA` managed certificates are valid only for 7 days.
To obtain a new certificate, deletion and re-creation of service binding are required.
The application also needs to be restarted to use the updated service binding.
Your call to the `XSUAA` to fetch a `JWT` will fail after your certificate expires.

:::
```

To extend the validity of the `XSUAA` managed certificate you could use parameters while creating a service binding.

`cf bind-service myapp myservice -c parameters.json`

```json
{
  "credential-type": "x509",
  "x509": {
    "key-length": 2048,
    "validity": 365,
    "validity-type": "DAYS"
  }
}
```
This way you can extend the validity of the generated `certificate` to a maximum of one year.

## Sticking to Client Credentials

> If you do not want to use Certificate based Authentication to obtain your `JWT`s and still want to continue using Client Credentials you can do that by configuring your service instances.
> Ensure that your `xs-security.json` contains the `instance-secret` as the first entry in property `credential-types`

Reviewer 3: ```suggestion
If you prefer to continue using **Client Credentials** instead of **Certificate-based Authentication**, make sure your `xs-security.json` contains the `instance-secret` as the first entry in the property `credential-types`. 
```

---
id: sdk-connectivity-mtls
title: Certificate-Based Authentication Using SAP Cloud SDK
hide_title: false
hide_table_of_contents: false
sidebar_label: Certificate Based Authentication
description: This article describes how the SAP Cloud SDK for Java can be used to establish connections to other cloud services using mTLS
keywords:
  - sap
  - cloud
  - sdk
  - mTLS
  - java
  - connectivity
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::danger

Currently, support for Certificate-based Authentication from SAP Cloud SDK is fully end-to-end tested only with the `XSUAA` service.
`Destination` and `Connectivity` services should also work but are not yet fully tested.

:::

The SAP Cloud SDK supports certificate-based authentication while establishing connections to other cloud services like `XSUAA` or `Destination` service.

## Why mTLS
Mutual `TLS` or `mTLS` for short is a method for mutual authentication.
It ensures that before proceeding with the `HTTP` exchange both the client and the server each mutually verify each other's identities by the use of `X.509` certificates.

## Usage of Client Credentials and Shortcomings

> To access a service protected by the  [`XSUAA`](https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/6373bb7a96114d619bfdfdc6f505d1b9.html) you'll need a `JWT`.

Reviewer 3: ```suggestion
Let's take an example of a service protected by the  [`XSUAA`](https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/6373bb7a96114d619bfdfdc6f505d1b9.html).
To access a service you'll need a `JWT`.
```

					"key": "key"
					...
				},
			}
		]
	}
}
```
This certificate would now be used to establish a secure session to obtain the `JWT` instead of the client credentials token flow used previously.
Please note that the process of obtaining the `JWT` is still internally carried out by the SAP Cloud SDK, we just use the `certificate` for this process now if the binding contains the `certificate`.

## Validity of Certificates
The `X.509` certificates used in the service binding could either be external or `XSUAA` managed.

:::caution

By default `XSUAA` managed certificates are only valid for 7 days and to obtain a new certificate, deletion and re-creation of service binding is required.
The application also needs to be restarted to update the service binding.

:::

To extend the validity of the `XSUAA` managed certificate you could use parameters while creating a service binding.

`cf bind-service myapp myservice -c parameters.json`

```json
{
  "credential-type": "x509",
  "x509": {
    "key-length": 2048,
    "validity": 365,
    "validity-type": "DAYS"
  }
}
```
This way you can extend the validity of the generated `certificate` to a maximum of one year.

> You can find more details on X.509 certificates [here](https://github.wdf.sap.corp/pages/CPSecurity/Knowledge-Base/03_ApplicationSecurity/X509Authentication/).

Reviewer 2: I don't know of any external documentation, so what can we do :/

> You can find more details on X.509 certificates [here](https://github.wdf.sap.corp/pages/CPSecurity/Knowledge-Base/03_ApplicationSecurity/X509Authentication/).

Reviewer: I'm not a fan of putting an internal link in our public docu. This is most certainly confusing for external customers... Maybe rather wait with this sentence until public documentation is available?

> You can find more details on X.509 certificates [here](https://github.wdf.sap.corp/pages/CPSecurity/Knowledge-Base/03_ApplicationSecurity/X509Authentication/).

Author: I will remove this sentence for now and add it back when we have a link to public documentation available.

}
```
This certificate would now be used to establish a secure session to obtain the `JWT` instead of the client credentials token flow used previously.

> Please note that the process of obtaining the `JWT` is still internally carried out by the SAP Cloud SDK.
> The only difference is that `certificate` is being used for this process now if the binding contains the `certificate`.

Reviewer: ```suggestion
Please note that the process of obtaining the `JWT` is still internally carried out by the SAP Cloud SDK.
The only difference is that the provided `certificate` is used to authenticate the client.
```
